pkg/cluster/acsk/action/action_utils.go

```diff
@@ -0,0 +1,345 @@
+	if err != nil {
+		if sdkErr, ok := err.(*aliErrors.ServerError); ok {
+			if strings.Contains(sdkErr.Message(), "ErrorClusterNotFound") {
+				// Cluster has been already deleted
+				return nil
+			}
+		}
+		return errors.WithMessage(err, fmt.Sprintf("DeleteClusterResponse: %#v \n", resp.BaseResponse))
+	}
+
+	if resp.GetHttpStatus() != http.StatusAccepted {
+		return fmt.Errorf("unexpected http status code: %d", resp.GetHttpStatus())
+	}
+
+	return nil
+}
+
+func describeScalingInstances(essClient *ess.Client, asgId, scalingConfId, regionId string) (*ess.DescribeScalingInstancesResponse, error) {
+	describeScalingInstancesRequest := ess.CreateDescribeScalingInstancesRequest()
+	describeScalingInstancesRequest.SetScheme(requests.HTTPS)
+	describeScalingInstancesRequest.SetDomain("ess." + regionId + ".aliyuncs.com")
+	describeScalingInstancesRequest.SetContentType(requests.Json)
+
+	describeScalingInstancesRequest.ScalingGroupId = asgId
+	describeScalingInstancesRequest.ScalingConfigurationId = scalingConfId
+
+	describeScalingInstancesResponse, err := essClient.DescribeScalingInstances(describeScalingInstancesRequest)
+	if err != nil {
+		return nil, emperror.WrapWith(err, "could not describe scaling instances", "scalingGroupId", asgId)
+	}
+	return describeScalingInstancesResponse, nil
+}
+
+func attachInstancesToCluster(log logrus.FieldLogger, clusterID string, instanceIds []string, csClient *cs.Client) (*acsk.AlibabaDescribeClusterResponse, error) {
+	log.Info("Attaching nodepools to cluster")
+	attachInstanceRequest := cs.CreateAttachInstancesRequest()
+	attachInstanceRequest.SetScheme(requests.HTTPS)
+	attachInstanceRequest.SetDomain(acsk.AlibabaApiDomain)
+	attachInstanceRequest.SetContentType(requests.Json)
+
+	attachInstanceRequest.ClusterId = clusterID
+
+	content := map[string]interface{}{
+		"instances": instanceIds,
+		"password":  "Hello1234", // Dummy password should be used here otherwise the api will fail
+	}
+	contentJSON, err := json.Marshal(content)
+	if err != nil {
+		return nil, err
+	}
+	attachInstanceRequest.SetContent(contentJSON)
+
+	_, err = csClient.AttachInstances(attachInstanceRequest)
+	if err != nil {
+		return nil, emperror.Wrap(err, "could not attach instances to cluster")
+	}
+	log.Info("Wait for nodepool attach")
+	clusterWithPools, err := waitUntilClusterCreateOrScaleComplete(log, clusterID, csClient, false)
+	if err != nil {
+		return nil, emperror.Wrap(err, "attaching nodepool failed")
+	}
+	return clusterWithPools, nil
+}
+
+func deleteNodepools(log logrus.FieldLogger, nodePools []*model.ACSKNodePoolModel, essClient *ess.Client, regionId string) error {
+	errChan := make(chan error, len(nodePools))
+	defer close(errChan)
+
+	for _, nodePool := range nodePools {
+		go func(nodePool *model.ACSKNodePoolModel) {
+
+			deleteSGRequest := ess.CreateDeleteScalingGroupRequest()
+			deleteSGRequest.SetScheme(requests.HTTPS)
+			deleteSGRequest.SetDomain("ess." + regionId + ".aliyuncs.com")
```

Comment:
The api endpoint could be extracted into a const similar to `ossEndpointFmt = "https://oss-%s.aliyuncs.com"`